sklearn/mixture/_gaussian_mixture_ic.py

```diff
@@ -474,7 +474,7 @@ def fit(self, X, y=None):
         self._check_parameters()
         X = check_array(
             X, dtype=[np.float64, np.float32], ensure_min_samples=1
-        )
+        ).copy()
```

Comment:
What is the `AssertionError` you are trying to fix with this change?

Reply:
I got this [AssertionError](https://github.com/scikit-learn/scikit-learn/pull/19562/checks?check_run_id=2511151020) and I tried to fix it using a `.copy()` inspired by this [code](https://github.com/scikit-learn/scikit-learn/blob/de1262c35e2aa4ee062d050281ee576ce9e35c94/sklearn/inspection/_permutation_importance.py#L26). Does it seem appropriate to you?

Reply:
Looking at the failing test case (copied here for posterity):
```
Mismatched elements: 33 / 50 (66%)
Max absolute difference: 1
Max relative difference: 1.
 x: array([1, 2, 1, 2, 0, 0, 0, 2, 0, 2, 1, 1, 2, 1, 2, 2, 0, 2, 2, 0, 1, 2,
       2, 1, 0, 1, 2, 0, 1, 2, 1, 0, 0, 1, 1, 1, 1, 2, 0, 2, 0, 0, 2, 0,
       2, 1, 0, 0, 0, 1], dtype=int64)
 y: array([2, 1, 2, 1, 0, 0, 0, 1, 0, 1, 2, 2, 1, 2, 1, 1, 0, 1, 1, 0, 2, 1,
       1, 2, 0, 2, 1, 0, 2, 1, 2, 0, 0, 2, 2, 2, 2, 1, 0, 1, 0, 0, 1, 0,
       1, 2, 0, 0, 0, 2], dtype=int64)
```

the results are identical up to a permutation (1 and 2 are swapped).
This is not much of a problem for clustering algorithms, but we still want them to assert the good centre labelling for points.

I don't know why this case happens, but even-though this patch solves the problem, we probably do not want to copy this user-provided array as it can be voluminous.

The best would be to dig into this case, understanding why the behaviour is different when `readonly_memmap=True`.

Reply:
@jjerphan Sorry for the late reply. Here is what I found. It appears to me that the two best models, among all models searched over by `GaussianMixtureIC` for this test, have the following parameters:
- `{'covariance_type': 'tied', 'linkage': 'ward', 'affinity': 'euclidean', 'n_components': 3}`
- `{'covariance_type': 'tied', 'linkage': 'complete', 'affinity': 'euclidean', 'n_components': 3}`

When fitting with `X.tolist()`, the BIC values of the two models are
> [105.20017905242764, 105.2001790524276]

When fitting with `X`, the BIC values of the two models are

> [105.20017905242759, 105.20017905242761]

Therefore, different models are selected in the two cases despite the tiny difference, leading to differential predictions (recreated [here](https://github.com/tliu68/scikit-learn/blob/gm_ic/GM_test.py)). (BTW, the difference somehow is gone when not using `memmap` or using `.copy()`)

In the newest [commit](https://github.com/scikit-learn/scikit-learn/pull/19562/commits/1da945df4587852c1c91aee7116e96b9e12f88ee), I changed the default to no Agglomerative initialization and fixed the [`reg_covar`](https://github.com/scikit-learn/scikit-learn/pull/19562#discussion_r624558117) variable (earlier I forgot to update the corresponding GM attribute) but now I'm seeing another [AssertionError](https://github.com/scikit-learn/scikit-learn/pull/19562/checks?check_run_id=2564646680). I'm still trying to replicate this error but do you think the current `reg_covar` temporary variable is added appropriately?
Thank you very much for your help!

Reply:
Hello @jjerphan ! I have updated the code so that the precision of BIC is changed based on sample size (a difference small than 10^{-10} for such a small dataset like above doesn't seem to be very meaningful), and a best model is selected randomly if there were a tie. It should take care of errors like above where a precision issue could cause a different model to be selected in a separate fit. Does that seem to be a better solution to you? Thank you for your help!

Reply:
Hi @tliu68, thanks for the report!
I will try to have a look at it soon.

In the meantime, the code formatting changed on `main` causing conflicts on this PR. 
Instructions given here https://github.com/scikit-learn/scikit-learn/issues/20301#issue-925041705 might help you solve the problem. :slightly_smiling_face:

Reply:
@jjerphan Please let me know if there is anything else I should modify! Thanks!

Reply:
Hi @tliu68

There's a conversation about your example left; I am going to mention you there.